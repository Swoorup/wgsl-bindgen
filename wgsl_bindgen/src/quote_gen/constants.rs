use proc_macro2::Ident;

/// This mod is used such that all the mods in the out can reference this from anywhere
pub(crate) const MOD_REFERENCE_ROOT: &str = "_root";
pub(crate) const MOD_STRUCT_ASSERTIONS: &str = "layout_asserts";
pub(crate) const MOD_BYTEMUCK_IMPLS: &str = "bytemuck_impls";

pub(crate) fn mod_reference_root() -> Ident {
  unsafe { syn::parse_str(MOD_REFERENCE_ROOT).unwrap_unchecked() }
}
