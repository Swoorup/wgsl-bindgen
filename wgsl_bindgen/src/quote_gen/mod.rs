mod constants;
mod rust_item;
mod rust_module_builder;
mod rust_struct_builder;
mod rust_type_info;

use core::panic;

pub(crate) use constants::*;
use proc_macro2::TokenStream;
pub(crate) use rust_item::*;
pub(crate) use rust_module_builder::*;
pub(crate) use rust_struct_builder::*;
pub(crate) use rust_type_info::*;

use crate::bevy_util::demangle_str;

/// Creates a raw string literal from the given shader content.
///
/// # Arguments
///
/// * `shader_content` - The content of the shader as a string.
///
/// # Returns
///
/// The token stream representing the raw string literal.
pub(crate) fn create_shader_raw_string_literal(shader_content: &str) -> TokenStream {
  syn::parse_str::<TokenStream>(&format!("r#\"\n{}\"#", &shader_content)).unwrap()
}

/// Demangles the given string and qualifies it with the qualification root.
///
/// # Arguments
///
/// * `string` - The string to demangle and qualify.
/// * `default_mod_path` - The default module path to use if the string does not contain a module path.
///
/// # Returns
///
/// The demangled and qualified token stream.
pub(crate) fn demangle_and_fully_qualify_str(
  string: &str,
  default_mod_path: Option<&str>,
) -> String {
  let demangled = demangle_str(string);

  match (demangled.contains("::"), default_mod_path) {
    (true, _) => {
      format!("{}::{}", MOD_REFERENCE_ROOT, demangled)
    }
    (false, None) => demangled.to_string(),
    (false, Some(default_mod_path)) => {
      if default_mod_path.is_empty() {
        panic!("default_mod_path cannot be empty");
      }

      let default_mod_path = default_mod_path.to_lowercase();
      format!("{MOD_REFERENCE_ROOT}::{default_mod_path}::{demangled}")
    }
  }
}

pub(crate) fn demangle_and_fully_qualify(
  string: &str,
  default_mod_path: Option<&str>,
) -> TokenStream {
  let raw_path = demangle_and_fully_qualify_str(string, default_mod_path);
  syn::parse_str(&raw_path).unwrap()
}

#[cfg(test)]
mod tests {
  use pretty_assertions::assert_eq;

  use super::demangle_and_fully_qualify;

  #[test]
  fn should_fully_qualify_mangled_string() {
    let string = "UniformsX_naga_oil_mod_XOR4XAZLTX";
    let actual = demangle_and_fully_qualify(string, None);
    assert_eq!(actual.to_string(), "_root :: types :: Uniforms");
  }

  #[test]
  fn should_not_fully_qualify_non_mangled_string() {
    let string = "MatricesF64";
    let actual = demangle_and_fully_qualify(string, None);
    assert_eq!(actual.to_string(), "MatricesF64");
  }
}
