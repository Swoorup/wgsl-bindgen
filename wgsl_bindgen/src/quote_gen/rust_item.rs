#![allow(dead_code)]
use derive_more::Constructor;
use proc_macro2::TokenStream;
use smol_str::SmolStr;

/// `RustItemPath` represents the path to a Rust item within a module.
#[derive(Constructor, Debug, Clone, PartialEq, Eq, Hash)]
pub(crate) struct RustItemPath {
  /// The path to the parent module.
  pub parent_module_path: SmolStr,
  /// name of the item, without the module path.
  pub item_name: SmolStr,
}

impl RustItemPath {
  pub fn get_fully_qualified_name(&self) -> SmolStr {
    if self.parent_module_path.is_empty() {
      SmolStr::new(self.item_name.as_str())
    } else {
      SmolStr::new(format!("{}::{}", self.parent_module_path, self.item_name).as_str())
    }
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub(crate) enum RustItemKind {
  ConstVarDecl,
  TraitImpls,
  Any,
}

/// Represents a Rust source item, that is either a ConstVar, TraitImpls or others.
#[derive(Constructor)]
pub(crate) struct RustItem {
  pub kind: RustItemKind,
  pub path: RustItemPath,
  pub item: TokenStream,
}
