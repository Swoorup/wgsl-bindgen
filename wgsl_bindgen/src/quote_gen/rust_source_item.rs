use proc_macro2::TokenStream;

use crate::bevy_util::demangle_splitting_mod_path_and_item;

/// Represents a Rust source item.
pub(crate) struct RustSourceItem {
  /// If not present this item belongs at the source root
  pub mod_path: Option<String>,
  pub name: String,
  pub item: TokenStream,
}

impl RustSourceItem {
  pub fn from_mangled(name: &str, item: TokenStream) -> Self {
    let (mod_path, name) = demangle_splitting_mod_path_and_item(name);

    Self {
      mod_path,
      name,
      item,
    }
  }
}
