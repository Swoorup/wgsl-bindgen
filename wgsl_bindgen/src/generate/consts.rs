use proc_macro2::Span;
use quote::quote;
use syn::Ident;

use crate::quote_gen::{RustItem, RustItemKind, RustItemPath};

pub fn consts_items(invoking_entry_module: &str, module: &naga::Module) -> Vec<RustItem> {
  // Create matching Rust constants for WGSl constants.
  module
    .constants
    .iter()
    .filter_map(|(_, t)| -> Option<RustItem> {
      let name_str = t.name.as_ref()?;

      // we don't need full qualification here
      let rust_item_path = RustItemPath::from_mangled(name_str, invoking_entry_module);
      let name = Ident::new(&rust_item_path.item_name, Span::call_site());

      // TODO: Add support for f64 and f16 once naga supports them.
      let type_and_value = match &module.global_expressions[t.init] {
        naga::Expression::Literal(literal) => match literal {
          naga::Literal::F64(v) => Some(quote!(f32 = #v)),
          naga::Literal::F32(v) => Some(quote!(f32 = #v)),
          naga::Literal::U32(v) => Some(quote!(u32 = #v)),
          naga::Literal::U64(v) => Some(quote!(u64 = #v)),
          naga::Literal::I32(v) => Some(quote!(i32 = #v)),
          naga::Literal::Bool(v) => Some(quote!(bool = #v)),
          naga::Literal::I64(v) => Some(quote!(i64 = #v)),
          naga::Literal::AbstractInt(v) => Some(quote!(i64 = #v)),
          naga::Literal::AbstractFloat(v) => Some(quote!(f64 = #v)),
        },
        _ => None,
      }?;

      Some(RustItem::new(
        RustItemKind::ConstVarDecl,
        rust_item_path,
        quote! { pub const #name: #type_and_value;},
      ))
    })
    .collect()
}

#[cfg(test)]
mod tests {
  use indoc::indoc;
  use proc_macro2::TokenStream;

  use super::*;
  use crate::assert_tokens_eq;

  pub fn consts(module: &naga::Module) -> Vec<TokenStream> {
    consts_items("", module)
      .into_iter()
      .map(|i| i.item)
      .collect()
  }

  #[test]
  fn write_global_consts() {
    let source = indoc! {r#"
            const INT_CONST = 12;
            const UNSIGNED_CONST = 34u;
            const FLOAT_CONST = 0.1;
            // TODO: Naga doesn't implement f16, even though it's in the WGSL spec
            // const SMALL_FLOAT_CONST:f16 = 0.1h;
            const BOOL_CONST = true;

            @fragment
            fn main() {
                // TODO: This is valid WGSL syntax, but naga doesn't support it apparently.
                // const C_INNER = 456;
            }
        "#};

    let module = naga::front::wgsl::parse_str(source).unwrap();

    let consts = consts(&module);
    let actual = quote!(#(#consts)*);
    eprintln!("{actual}");

    assert_tokens_eq!(
      quote! {
          pub const INT_CONST: i32 = 12i32;
          pub const UNSIGNED_CONST: u32 = 34u32;
          pub const FLOAT_CONST: f32 = 0.1f32;
          pub const BOOL_CONST: bool = true;
      },
      actual
    );
  }
}
