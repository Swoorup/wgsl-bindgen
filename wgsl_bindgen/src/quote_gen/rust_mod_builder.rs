use miette::Diagnostic;
use naga::FastIndexMap;
use proc_macro2::TokenStream;
use quote::quote;
use smallvec::SmallVec;
use syn::Ident;
use thiserror::Error;

use super::{constants::MOD_REFERENCE_ROOT, RustSourceItem};
use crate::quote_gen::constants::mod_reference_root;

#[derive(Debug, Error, Diagnostic)]
pub enum RustModBuilderError {
  #[error("Different Content for unique id {id}")]
  DuplicateContentError {
    id: String,
    existing: String,
    received: String,
  },
}

#[derive(Default)]
struct RustMod {
  name: String,
  is_public: bool,
  module_attributes: TokenStream,
  initial_contents: TokenStream,
  content: Vec<TokenStream>,
  unique_content: FastIndexMap<String, usize>,
  submodules: FastIndexMap<String, RustMod>,
}

impl RustMod {
  fn new(name: &str, is_public_visibility: bool, initial_contents: TokenStream) -> Self {
    Self {
      module_attributes: quote!(),
      name: name.to_owned(),
      is_public: is_public_visibility,
      initial_contents,
      content: Vec::new(),
      unique_content: FastIndexMap::default(),
      submodules: FastIndexMap::default(),
    }
  }

  fn add_content(&mut self, content: TokenStream) {
    self.content.push(content);
  }

  fn add_unique(
    &mut self,
    id: &str,
    content: TokenStream,
  ) -> Result<(), RustModBuilderError> {
    if let Some(existing_content) = self
      .unique_content
      .get(id)
      .and_then(|&index| self.content.get(index))
    {
      let existing = existing_content.to_string();
      let received = content.to_string();

      if existing != received {
        return Err(RustModBuilderError::DuplicateContentError {
          id: id.to_string(),
          existing,
          received,
        });
      }
    } else {
      self
        .unique_content
        .insert(id.to_string(), self.content.len());
      self.content.push(content);
    }

    Ok(())
  }

  fn get_or_create_submodule(&mut self, name: &str) -> &mut RustMod {
    self
      .submodules
      .entry(name.to_owned())
      .or_insert_with(|| RustMod::new(name, true, self.initial_contents.clone()))
  }

  fn generate(&self) -> TokenStream {
    let name = Ident::new(&self.name, proc_macro2::Span::call_site());

    let initial_contents = &self.initial_contents;
    let content = &self.content;

    let visibility = if self.is_public {
      quote!(pub)
    } else {
      quote!()
    };

    let submodules = self
      .submodules
      .values()
      .map(|m| m.generate())
      .collect::<Vec<_>>();

    let mod_attr = &self.module_attributes;

    quote! {
      #mod_attr
      #visibility mod #name {
          #initial_contents
          #( #content )*
          #( #submodules )*
      }
    }
  }
}

#[derive(Clone, Copy)]
pub struct RustModBuilderConfig {
  use_relative_root: bool,
}

impl RustModBuilderConfig {
  fn build_module(&self, mod_name: &str) -> RustMod {
    if self.use_relative_root {
      // this helps import relative items for nested mods under this root
      // https://discord.com/channels/442252698964721669/448238009733742612/1207323647203868712
      let root = mod_reference_root();
      if mod_name == MOD_REFERENCE_ROOT {
        RustMod {
          name: mod_name.into(),
          is_public: false,
          module_attributes: quote!(#[allow(unused)]),
          initial_contents: quote! {pub use super::*;},
          ..Default::default()
        }
      } else {
        RustMod {
          name: mod_name.into(),
          is_public: true,
          module_attributes: quote!(),
          initial_contents: quote! {
            #[allow(unused_imports)]
            use super::{#root, #root::*};
          },
          ..Default::default()
        }
      }
    } else {
      RustMod::new(mod_name, true, quote!())
    }
  }

  fn initial_modules(&self) -> FastIndexMap<String, RustMod> {
    if self.use_relative_root {
      let name = MOD_REFERENCE_ROOT.to_owned();
      let root_mod = self.build_module(name.as_str());
      FastIndexMap::from_iter([(name, root_mod)])
    } else {
      Default::default()
    }
  }
}

pub(crate) struct RustModBuilder {
  modules: FastIndexMap<String, RustMod>,
  config: RustModBuilderConfig,
}

impl RustModBuilder {
  pub fn new(use_relative_root: bool) -> Self {
    let config = RustModBuilderConfig { use_relative_root };

    Self {
      modules: config.initial_modules(),
      config,
    }
  }

  fn get_or_create_module(&mut self, path: &str) -> &mut RustMod {
    if path.is_empty() {
      panic!("path cannot be empty");
    }

    let modules = path.split("::").collect::<SmallVec<[_; 8]>>();

    let mut current_module = self
      .modules
      .entry(modules[0].to_owned())
      .or_insert_with(|| self.config.build_module(modules[0]));

    for name in &modules[1..] {
      current_module = current_module.get_or_create_submodule(name);
    }
    current_module
  }

  pub fn add_items(
    &mut self,
    default_mod_path: &str,
    items: Vec<RustSourceItem>,
  ) -> Result<(), RustModBuilderError> {
    for item in items {
      let module_path = item
        .mod_path
        .as_ref()
        .map(|s| s.as_str())
        .unwrap_or(default_mod_path);

      self.add_unique(module_path, &item.name, item.item)?;
    }

    Ok(())
  }

  pub fn add(&mut self, path: &str, content: TokenStream) {
    self.get_or_create_module(path).add_content(content);
  }

  pub fn add_unique(
    &mut self,
    path: &str,
    id: &str,
    content: TokenStream,
  ) -> Result<(), RustModBuilderError> {
    self.get_or_create_module(path).add_unique(id, content)
  }

  /// Generates the top level root module that includes other modules
  pub fn generate(&self) -> TokenStream {
    let modules: Vec<TokenStream> = self.modules.values().map(|m| m.generate()).collect();
    quote! {
      #( #modules )*
    }
  }
}

#[cfg(test)]
mod tests {
  use pretty_assertions::assert_eq;
  use quote::quote;

  use super::{RustModBuilder, RustModBuilderError};
  use crate::assert_tokens_eq;

  #[test]
  fn test_module_generation_works() {
    let mut mod_builder = RustModBuilder::new(false);
    mod_builder.add("a::b::c::d", quote! {struct A;});
    mod_builder.add("a::b::c", quote! {struct B;});
    mod_builder.add("a::b::c", quote! {struct C;});

    let actual = mod_builder.generate();

    assert_tokens_eq!(
      actual,
      quote! {
        pub mod a {
          pub mod b {
            pub mod c {
              struct B ;
              struct C ;
              pub mod d {
                struct A ;
              }
            }
          }
        }
      }
    );
  }

  #[test]
  fn test_relative_root_feature() {
    let mut mod_builder = RustModBuilder::new(true);
    mod_builder.add("a::b", quote! {struct A;});
    mod_builder.add(
      "a",
      quote! {struct B{
        a: a::b::A
      }},
    );

    let actual = mod_builder.generate();

    assert_tokens_eq!(
      actual,
      quote! {
        #[allow(unused)]
        mod _root {
          pub use super::*;
        }
        pub mod a {
          #[allow(unused_imports)]
          use super::{_root, _root::*};
          struct B {
              a: a::b::A,
          }
          pub mod b {
              #[allow(unused_imports)]
              use super::{_root, _root::*};
              struct A;
          }
        }
      }
    );
  }

  #[test]
  fn test_module_add_duplicates() -> Result<(), RustModBuilderError> {
    let mut mod_builder = RustModBuilder::new(false);
    mod_builder.add_unique("a::b", "A", quote! {struct A;})?;
    mod_builder.add_unique("a", "A", quote! {struct B;})?;
    mod_builder.add_unique("a::b", "A", quote! {struct A;})?;

    let tokens = mod_builder.generate();

    assert_tokens_eq!(
      tokens,
      quote! {
        pub mod a {
          struct B;
          pub mod b {
            struct A ;
          }
        }
      }
    );
    Ok(())
  }

  #[test]
  fn test_module_add_duplicates_different_contents() {
    let mut mod_builder = RustModBuilder::new(false);
    mod_builder
      .add_unique("a::b", "A", quote! {struct A;})
      .unwrap();

    let error = mod_builder.add_unique("a::b", "A", quote! {struct B;});

    assert_eq!(error.is_err(), true);
  }
}
