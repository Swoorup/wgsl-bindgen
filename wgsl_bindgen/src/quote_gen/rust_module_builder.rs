#![allow(unused)]

use miette::Diagnostic;
use proc_macro2::TokenStream;
use quote::quote;
use smallvec::SmallVec;
use syn::Ident;
use thiserror::Error;

use super::constants::MOD_REFERENCE_ROOT;
use super::{RustItem, RustItemKind};
use crate::quote_gen::constants::mod_reference_root;
use crate::FastIndexMap;

#[derive(Debug, Error, Diagnostic)]
pub enum RustModuleBuilderError {
  #[error("Different Content for unique id {id}")]
  DuplicateContentError {
    id: String,
    existing: String,
    received: String,
  },
}

struct UniqueItemInfo {
  index: usize,
  kind: RustItemKind,
}

#[derive(Default)]
struct RustModule {
  name: String,
  is_public: bool,
  module_attributes: TokenStream,
  initial_contents: TokenStream,
  content: Vec<TokenStream>,
  unique_content_info: FastIndexMap<String, UniqueItemInfo>,
  submodules: FastIndexMap<String, RustModule>,
}

impl RustModule {
  fn new(name: &str, is_public_visibility: bool, initial_contents: TokenStream) -> Self {
    Self {
      module_attributes: quote!(),
      name: name.to_owned(),
      is_public: is_public_visibility,
      initial_contents,
      content: Vec::new(),
      unique_content_info: FastIndexMap::default(),
      submodules: FastIndexMap::default(),
    }
  }

  fn add_content(&mut self, content: TokenStream) {
    self.content.push(content);
  }

  /// Adds unique content to the `RustModule`.
  ///
  /// This function checks if the provided `id` already exists in the `unique_content_info` map.
  /// If it does, it compares the existing content with the new content.
  /// If the content is the same, it does nothing.
  /// If the content is different but the `RustItemKind` is the same, it appends the new content to the existing content.
  /// If the `RustItemKind` is different, it returns an error.
  fn add_unique(
    &mut self,
    id: &str,
    kind: RustItemKind,
    content: TokenStream,
  ) -> Result<(), RustModuleBuilderError> {
    if let Some((prev_kind, existing_content)) =
      self.unique_content_info.get(id).and_then(|info| {
        let content = self.content.get_mut(info.index)?;
        Some((info.kind, content))
      })
    {
      if prev_kind == kind {
        let existing = existing_content.to_string();
        let received = content.to_string();
        if existing != received {
          return Err(RustModuleBuilderError::DuplicateContentError {
            id: id.to_string(),
            existing,
            received,
          });
        }
      } else {
        existing_content.extend(content);
      }
    } else {
      self.unique_content_info.insert(
        id.to_string(),
        UniqueItemInfo {
          index: self.content.len(),
          kind,
        },
      );
      self.content.push(content);
    }

    Ok(())
  }

  fn get_or_create_submodule(&mut self, name: &str) -> &mut RustModule {
    self
      .submodules
      .entry(name.to_owned())
      .or_insert_with(|| RustModule::new(name, true, self.initial_contents.clone()))
  }

  fn merge(&mut self, other: Self) {
    self.content.extend(other.content);
    self.unique_content_info.extend(other.unique_content_info);
    for (name, other_submodule) in other.submodules {
      let self_submodule = self.get_or_create_submodule(&name);
      self_submodule.merge(other_submodule);
    }
  }

  fn generate(&self) -> TokenStream {
    let name = Ident::new(&self.name, proc_macro2::Span::call_site());

    let initial_contents = &self.initial_contents;
    let content = &self.content;

    let visibility = if self.is_public {
      quote!(pub)
    } else {
      quote!()
    };

    let submodules = self
      .submodules
      .values()
      .map(|m| m.generate())
      .collect::<Vec<_>>();

    let mod_attr = &self.module_attributes;

    quote! {
      #mod_attr
      #visibility mod #name {
          #initial_contents
          #( #content )*
          #( #submodules )*
      }
    }
  }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct RustModBuilderConfig {
  use_relative_root: bool,
  generate_relative_root: bool,
}

impl RustModBuilderConfig {
  fn build_module(&self, mod_name: &str) -> RustModule {
    if self.use_relative_root {
      // this helps import relative items for nested mods under this root
      // https://discord.com/channels/442252698964721669/448238009733742612/1207323647203868712
      let root = mod_reference_root();
      if mod_name == MOD_REFERENCE_ROOT {
        RustModule {
          name: mod_name.into(),
          is_public: false,
          module_attributes: quote!(),
          initial_contents: quote! {pub use super::*;},
          ..Default::default()
        }
      } else {
        RustModule {
          name: mod_name.into(),
          is_public: true,
          module_attributes: quote!(),
          initial_contents: quote! {
            use super::{#root, #root::*};
          },
          ..Default::default()
        }
      }
    } else {
      RustModule::new(mod_name, true, quote!())
    }
  }

  fn initial_modules(&self) -> FastIndexMap<String, RustModule> {
    if self.use_relative_root && self.generate_relative_root {
      let name = MOD_REFERENCE_ROOT.to_owned();
      let root_mod = self.build_module(name.as_str());
      FastIndexMap::from_iter([(name, root_mod)])
    } else {
      Default::default()
    }
  }
}

pub(crate) struct RustModBuilder {
  modules: FastIndexMap<String, RustModule>,
  config: RustModBuilderConfig,
}

impl RustModBuilder {
  pub fn new(use_relative_root: bool, generate_relative_root: bool) -> Self {
    let config = RustModBuilderConfig {
      use_relative_root,
      generate_relative_root,
    };

    Self {
      modules: config.initial_modules(),
      config,
    }
  }

  fn get_or_create_module(&mut self, path: &str) -> &mut RustModule {
    if path.is_empty() {
      panic!("path cannot be empty");
    }

    let modules = path.split("::").collect::<SmallVec<[_; 8]>>();

    let mut current_module = self
      .modules
      .entry(modules[0].to_owned())
      .or_insert_with(|| self.config.build_module(modules[0]));

    for name in &modules[1..] {
      current_module = current_module.get_or_create_submodule(name);
    }
    current_module
  }

  pub fn add_items(
    &mut self,
    items: Vec<RustItem>,
  ) -> Result<(), RustModuleBuilderError> {
    for item in items {
      let module_path = item.path.module;
      let name = item.path.name;

      let mut m = self.get_or_create_module(&module_path);
      m.add_unique(&name, item.kind, item.item)?;
    }

    Ok(())
  }

  pub fn add(&mut self, path: &str, content: TokenStream) {
    self.get_or_create_module(path).add_content(content);
  }

  fn add_unique(
    &mut self,
    path: &str,
    id: &str,
    content: TokenStream,
  ) -> Result<(), RustModuleBuilderError> {
    self
      .get_or_create_module(path)
      .add_unique(id, RustItemKind::Any, content)
  }

  pub fn merge(mut self, other: Self) -> Self {
    assert_eq!(self.config, other.config);
    for (name, other_module) in other.modules {
      let self_module = self.get_or_create_module(&name);
      self_module.merge(other_module);
    }
    self
  }

  /// Generates the top level root module that includes other modules
  pub fn generate(&self) -> TokenStream {
    let modules: Vec<TokenStream> = self.modules.values().map(|m| m.generate()).collect();
    quote! {
      #( #modules )*
    }
  }
}

#[cfg(test)]
mod tests {
  use pretty_assertions::assert_eq;
  use quote::quote;

  use super::{RustModBuilder, RustModuleBuilderError};
  use crate::assert_tokens_eq;

  #[test]
  fn test_module_generation_works() {
    let mut mod_builder = RustModBuilder::new(false, false);
    mod_builder.add("a::b::c::d", quote! {struct A;});
    mod_builder.add("a::b::c", quote! {struct B;});
    mod_builder.add("a::b::c", quote! {struct C;});

    let actual = mod_builder.generate();

    assert_tokens_eq!(
      actual,
      quote! {
        pub mod a {
          pub mod b {
            pub mod c {
              struct B ;
              struct C ;
              pub mod d {
                struct A ;
              }
            }
          }
        }
      }
    );
  }

  #[test]
  fn test_relative_root_feature() {
    let mut mod_builder = RustModBuilder::new(true, true);
    mod_builder.add("a::b", quote! {struct A;});
    mod_builder.add(
      "a",
      quote! {struct B{
        a: a::b::A
      }},
    );

    let actual = mod_builder.generate();

    assert_tokens_eq!(
      actual,
      quote! {
        mod _root {
          pub use super::*;
        }
        pub mod a {
          use super::{_root, _root::*};
          struct B {
              a: a::b::A,
          }
          pub mod b {
              use super::{_root, _root::*};
              struct A;
          }
        }
      }
    );
  }

  #[test]
  fn test_include_relative_root_but_dont_generate_it() {
    let mut mod_builder = RustModBuilder::new(true, false);
    mod_builder.add("a::b", quote! {struct A;});
    mod_builder.add(
      "a",
      quote! {struct B{
        a: a::b::A
      }},
    );

    let actual = mod_builder.generate();

    assert_tokens_eq!(
      actual,
      quote! {
        pub mod a {
          use super::{_root, _root::*};
          struct B {
              a: a::b::A,
          }
          pub mod b {
              use super::{_root, _root::*};
              struct A;
          }
        }
      }
    );
  }

  #[test]
  fn test_module_add_duplicates() -> Result<(), RustModuleBuilderError> {
    let mut mod_builder = RustModBuilder::new(false, false);
    mod_builder.add_unique("a::b", "A", quote! {struct A;})?;
    mod_builder.add_unique("a", "A", quote! {struct B;})?;
    mod_builder.add_unique("a::b", "A", quote! {struct A;})?;

    let tokens = mod_builder.generate();

    assert_tokens_eq!(
      tokens,
      quote! {
        pub mod a {
          struct B;
          pub mod b {
            struct A ;
          }
        }
      }
    );
    Ok(())
  }

  #[test]
  fn test_module_add_duplicates_different_contents() {
    let mut mod_builder = RustModBuilder::new(false, false);
    mod_builder
      .add_unique("a::b", "A", quote! {struct A;})
      .unwrap();

    let error = mod_builder.add_unique("a::b", "A", quote! {struct B;});

    assert_eq!(error.is_err(), true);
  }

  #[test]
  fn test_merge() {
    let mut builder1 = RustModBuilder::new(false, false);
    builder1.add("a::b::c", quote! {struct A;});
    builder1.add("a::b::d", quote! {struct B;});

    let mut builder2 = RustModBuilder::new(false, false);
    builder2.add("a::b::c", quote! {struct C;});
    builder2.add("a::b::e", quote! {struct D;});

    let actual = builder1.merge(builder2).generate();

    assert_tokens_eq!(
      actual,
      quote! {
        pub mod a {
          pub mod b {
            pub mod c {
              struct A;
              struct C;
            }
            pub mod d {
              struct B;
            }
            pub mod e {
              struct D;
            }
          }
        }
      }
    );
  }
}
