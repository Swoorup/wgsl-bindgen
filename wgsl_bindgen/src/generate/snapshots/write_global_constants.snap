---
source: wgsl_bindgen/src/generate/consts.rs
---
pub const UNSIGNED_CONST: u32 = 34u32;
pub const SMALL_FLOAT_CONST: half::f16 = half::f16::from_f32_const(0.099975586f32);
pub const BOOL_CONST: bool = true;
